Nebula

jeanyu-habana opened the pull request:
## Description

support NebulaGraph cluster deployment in k8s

## Issues

n/a

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [ ] Others (enhancement, documentation, validation, etc.)

## Dependencies

n/a

## Tests

Tested the deployment and config scripts. NebulaGraph services were up and running. DB spaces can be created and managed.

letonghan commented on comps/third_parties/neo4j/deployment/kubernetes/README.md:
```diff
@@ -0,0 +1,17 @@
+# Deploy Neo4j on kubernetes cluster
```
Hi, these four files in `neo4j/deployment/kubernetes` seems the same as in [PR1254](https://github.com/opea-project/GenAIComps/pull/1254). Will you delete one of them before merged?

jeanyu-habana replied:
deleted the 4 extras from this PR. @letonghan

lvliang-intel commented on comps/third_parties/nebula/deployment/kubernetes/README.md:
```diff
@@ -0,0 +1,25 @@
+# Deploy NebulaGraph cluster (with Kubectl)
```
Could you please add a description of Nebula and instructions on how to access the service?

jeanyu-habana replied:
added info for accessing the service (aka the databases).

letonghan commented:
Hi @jeanyu-habana , Please check the failed `Online Document Building` CI.
Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!

jeanyu-habana commented:
> Hi @jeanyu-habana , Please check the failed `Online Document Building` CI. Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!

Added component test script and fixed document building CI error. Please let me know if there are additional comments before approval. Thanks! @letonghan

chensuyue commented:
> > Hi @jeanyu-habana , Please check the failed `Online Document Building` CI. Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!
> 
> Added component test script and fixed document building CI error. Please review.
> 
> Do you know if there is a document with information on the target k8s cluster? For example, storageClass, namespace, etc. Also, what's the general process for adding docker images for our services to OPEA's docker hub: https://hub.docker.com/u/opea? These questions are for my future PRs. Thanks! @letonghan

If you want to release a new image into docker hub, add the image information [here](https://github.com/opea-project/GenAIComps/tree/main/.github/workflows/docker/compose), and it will be released on docker hub in the latest release schedule. The latest future release version is v1.3, which will be released on 4/25.

chensuyue commented:
> > Hi @jeanyu-habana , Please check the failed `Online Document Building` CI. Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!
> 
> Added component test script and fixed document building CI error. Please review.
> 
> Do you know if there is a document with information on the target k8s cluster? For example, storageClass, namespace, etc. Also, what's the general process for adding docker images for our services to OPEA's docker hub: https://hub.docker.com/u/opea? These questions are for my future PRs. Thanks! @letonghan

For k8s deployment, refer to `comps/xxx/deployment/kubernetes/README.md`, like this one, https://github.com/opea-project/GenAIComps/blob/main/comps/asr/deployment/kubernetes/README.md. The template itself is in https://github.com/opea-project/GenAIInfra/tree/main/helm-charts/common. When you add a new values in GenAIComps and the template is available in Github Package, then the CI will be triggered automatically on our k8s cluster.

jeanyu-habana commented:
> > Hi @jeanyu-habana , Please check the failed `Online Document Building` CI. Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!
> 
> Added component test script and fixed document building CI error. Please review.
> 
> Do you know if there is a document with information on the target k8s cluster? For example, storageClass, namespace, etc. Also, what's the general process for adding docker images for our services to OPEA's docker hub: https://hub.docker.com/u/opea? These questions are for my future PRs. Thanks! @letonghan

Would you be able to review the changes sometime soon? Any additional comments before the issues can be resolved? Would love to see the PR merged soon. Please let me know. Thanks! @letonghan

jeanyu-habana commented:
> > > Hi @jeanyu-habana , Please check the failed `Online Document Building` CI. Also, please add a test script for your third-party component [here](https://github.com/opea-project/GenAIComps/tree/main/tests/third_parties), following the naming schema. Thanks!
> > 
> > 
> > Added component test script and fixed document building CI error. Please review.
> > Do you know if there is a document with information on the target k8s cluster? For example, storageClass, namespace, etc. Also, what's the general process for adding docker images for our services to OPEA's docker hub: https://hub.docker.com/u/opea? These questions are for my future PRs. Thanks! @letonghan
> 
> For k8s deployment, refer to `comps/xxx/deployment/kubernetes/README.md`, like this one, https://github.com/opea-project/GenAIComps/blob/main/comps/asr/deployment/kubernetes/README.md. The template itself is in https://github.com/opea-project/GenAIInfra/tree/main/helm-charts/common. When you add a new values in GenAIComps and the template is available in Github Package, then the CI will be triggered automatically on our k8s cluster.

Thank you for your prompt responses. I appreciate the help. @chensuyue

letonghan reviewed (approved):
lgtm

xiguiw commented:
@lianhao @yongfengdu 
Would you please review this PR? Does the structure, style etc. confirm to ours?

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/README.md:
````diff
@@ -0,0 +1,64 @@
+```bash
+helm repo add nebula-operator https://vesoft-inc.github.io/nebula-operator/charts
+helm repo update
+helm install nebula-operator nebula-operator/nebula-operator --namespace=<namespace_name> --version=${chart_version}
````
We should either give specific chart version or let user to use the latest.

jeanyu-habana replied:
Okay, updated with a specific version.

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/README.md:
````diff
@@ -0,0 +1,64 @@
+## 1. Install cert-manager
+
+```bash
+kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.9.1/cert-manager.yaml
````
Can we have cert-manager of versions othern than 1.9.1? I'm asking this because the user might have other versions of cert-manager in their cluster already.

jeanyu-habana replied:
good catch. I replaced specific version with <release_version> in the README to indicate that the user can use other versions. Thanks.

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/README.md:
````diff
@@ -0,0 +1,64 @@
+Run the following command to check the IP of the Service:
+
+```bash
+$ kubectl get service -l app.kubernetes.io/cluster=<nebula>  #<nebula> is a variable value. Replace it with the desired name.
````
what should <nebula> be? Replace it by what value?

jeanyu-habana replied:
This can be any desired name.

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/README.md:
````diff
@@ -0,0 +1,64 @@
+Run the following command to connect to the NebulaGraph database using the IP of the <cluster-name>-graphd-svc Service above:
+
+```bash
+kubectl run -ti --image vesoft/nebula-console --restart=Never -- <nebula_console_name> -addr <cluster_ip>  -port <service_port> -u <username> -p <password>
````
It'd be better to give users direct bash commands to get <cluster_ip>, <service_port> .

jeanyu-habana replied:
Added the bash commands to find <cluster_ip>, <service_port> in README.

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/community_edition.yaml:
```diff
@@ -0,0 +1,75 @@
+      resources:
+        requests:
+          storage: 2Gi
+      storageClassName: "${STORAGE_CLASS_NAME}"
```
If nebula requires k8s to have a valid storage class, we'd better to say it clearly in README. Also if we follow the README strictly, this `${STORAGE_CLASS_NAME}` dereference could cause the `kubectl create -f enterprise_edition.yaml` fail due to the mismatch of storageClassName. Does nebula operator support not setting specific storageClassName?

lianhao commented on comps/third_parties/nebula/deployment/kubernetes/community_edition.yaml:
```diff
@@ -0,0 +1,75 @@
+        cpu: "1"
+        memory: "1Gi"
+    replicas: 1
+    image: vesoft/nebula-graphd
```
should we pin the image tag instead of using latest image? Just in case future versions of nebula might not be compatible with this resource operator definition.

yongfengdu commented:
I would propose to add this (If necessary) together with you mentioned HybridRAG.
Otherwise, it's not clear why we need to add this components - Adding an open source GraphDB and use OEPA CI to test it.

xiguiw commented:
@jeanyu-habana 

There is a graphRAG component implemented by @rbrugaro 

https://github.com/opea-project/GenAIComps/blob/c70f86891a7730f1529dba885b8b7826acf22cb2/comps/dataprep/src/integrations/neo4j_llamaindex.py#L82 

It is integrated into data-prep, to improve the OPEA RAG through data-prep and retriever.

To bring the NebulaGraph in, I have some questions.

1. What's the difference (or gap) between NebulaGraph  and the existed [GraphRAG](https://github.com/opea-project/GenAIComps/blob/c70f86891a7730f1529dba885b8b7826acf22cb2/comps/dataprep/src/integrations/neo4j_llamaindex.py#L82)?

2. How the NebulaGraph being integrated into OPEA components or improve the OPEA components functionality? 
Build an example with third party is not the targets of OPEA because there is no contribution from OPEA.
Could you integrate it with OPEA retriever/data-prep?

Thanks!

xiguiw commented:
@jeanyu-habana 
Would you please address the comments and questions?  Thanks!

yongfengdu commented:
@jeanyu-habana I was convinced that, in a sense of providing more components for OPEA project, so any users could utilize this components for new AI apps, it's ok to put these instructions for reference.
Because we don't have related helm charts for this in https://github.com/opea-project/GenAIInfra, the test logic would be different and you'll need to maintain it by yourself.
> I would propose to add this (If necessary) together with you mentioned HybridRAG. Otherwise, it's not clear why we need to add this components - Adding an open source GraphDB and use OEPA CI to test it.

jeanyu-habana commented:
> @jeanyu-habana
> 
> There is a graphRAG component implemented by @rbrugaro
> 
> https://github.com/opea-project/GenAIComps/blob/c70f86891a7730f1529dba885b8b7826acf22cb2/comps/dataprep/src/integrations/neo4j_llamaindex.py#L82
> 
> It is integrated into data-prep, to improve the OPEA RAG through data-prep and retriever.
> 
> To bring the NebulaGraph in, I have some questions.
> 
> 1. What's the difference (or gap) between NebulaGraph  and the existed [GraphRAG](https://github.com/opea-project/GenAIComps/blob/c70f86891a7730f1529dba885b8b7826acf22cb2/comps/dataprep/src/integrations/neo4j_llamaindex.py#L82)?
> 2. How the NebulaGraph being integrated into OPEA components or improve the OPEA components functionality?
>    Build an example with third party is not the targets of OPEA because there is no contribution from OPEA.
>    Could you integrate it with OPEA retriever/data-prep?
> 
> Thanks!

1. Existing GraphRAG in OPEA uses Neo4J graph database. Several of RAG related microservices we are about to submit to OPEA can use either Neo4J or NebulaGraph. NebulaGraph is distrubuted and much faster than Neo4J, especially during query and used in production environment. The difference between Neo4J and NebulaGraph can be found here: https://www.nebula-graph.io/posts/Neo4j_Alternative_Open-Source_Distributed_and_Lightning_Fast. 
2. There are multi ways we can integrate NebulaGraph microservice into the new projects we are about to submit. We can also integrate it into the existing OPEA retriever/data-prep (as you suggested above).

jeanyu-habana commented:
> @jeanyu-habana I was convinced that, in a sense of providing more components for OPEA project, so any users could utilize this components for new AI apps, it's ok to put these instructions for reference. Because we don't have related helm charts for this in https://github.com/opea-project/GenAIInfra, the test logic would be different and you'll need to maintain it by yourself.
> 
> > I would propose to add this (If necessary) together with you mentioned HybridRAG. Otherwise, it's not clear why we need to add this components - Adding an open source GraphDB and use OEPA CI to test it.

Agreed. Thank you. We will make sure at least one of our submissions will use Nebula and have OPEA CI test enabled.

jeanyu-habana commented:
> @jeanyu-habana Would you please address the comments and questions? Thanks!

Please see "quoted" reply in the context of your original question.

jeanyu-habana commented:
@lianhao Could you please review the changes I made per your requests? Thank you.

joshuayao reviewed (changes requested):
Thank you for your contribution! As per the OPEA contribution guidelines, new components are required to include Docker Compose deployment support. Could you please provide this to ensure compliance with the guidelines?

xiguiw commented:
> 2\. There are multi ways we can integrate NebulaGraph microservice into the new projects we are about to submit. We can also integrate it into the existing OPEA retriever/data-prep (as you suggested above).

Only Data-pre and retrieval interacts with graph DB. Maybe you can refer to Graph RAG to integrate NebulaGraph into data-prep/retrieval.

lvliang-intel commented on tests/third_parties/test_third_parties_nebula.sh:
```diff
@@ -0,0 +1,118 @@
+    echo "[ test create ] creating space.."
+    query="CREATE SPACE my_space(partition_num=10, replica_factor=1, vid_type=FIXED_STRING(32)); USE my_space; CREATE TAG person(name string, age int);"
+
+    create_response=$(kubectl run -ti --rm --image vesoft/nebula-console --restart=Never -- nebula-console -addr "$cluster_ip" -port 9669 -u root -p vesoft -e "$query" 2>&1)
```
How can other microservices access Nebula Database? Are there any software APIs or endpoints available? Using Nebula-Console is not good way for this purpose.

xiguiw replied:
> How can other microservices access Nebula Database? Are there any software APIs or endpoints available? Using Nebula-Console is not good way for this purpose.

@jeanyu-habana 

Could you resolve this?
